helix-core/src/comment.rs

```diff
@@ -94,6 +94,168 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+                commented = false;
+            } else if commented {
+                to_change.push((*selection, open_pos, close_pos, open_margin, close_margin));
+            }
+        }
+    }
+    if to_change.is_empty() {
+        commented = false;
+    }
+    (commented, to_change)
+}
+
+#[must_use]
+pub fn toggle_block_comments(
+    doc: &Rope,
+    selection: &Selection,
+    tokens: Option<(&str, &str)>,
+) -> Transaction {
+    let (open_token, close_token) = tokens.unwrap_or(("/*", "*/"));
+    let text = doc.slice(..);
+    let (commented, to_change) = find_block_comments(open_token, close_token, text, selection);
+    let open = Tendril::from(format!("{} ", open_token));
+    let close = Tendril::from(format!(" {}", close_token));
+    let mut changes: Vec<Change> = Vec::with_capacity(selection.len());
+    for (range, open_pos, close_pos, open_margin, close_margin) in to_change {
+        let from = range
+            .with_direction(crate::movement::Direction::Forward)
+            .from();
+        if commented {
+            changes.push((
+                from + open_pos,
+                from + open_pos + open_token.len() + open_margin,
+                None,
+            ));
+            changes.push((
+                from + close_pos - close_token.len() - close_margin + 1,
+                from + close_pos + 1,
+                None,
+            ));
+        } else {
+            changes.push((from + open_pos, from + open_pos, Some(open.clone())));
+            changes.push((
+                from + close_pos + 1,
+                from + close_pos + 1,
+                Some(close.clone()),
+            ));
+        }
+    }
+    Transaction::change(doc, changes.into_iter())
+}
+
+/// Return if the selections should be block or line comment toggled
+pub fn comment_type(
+    token: Option<&str>,
+    tokens: Option<(&str, &str)>,
+    text: RopeSlice,
+    selection: &Selection,
+) -> bool {
+    let mut lines: Vec<usize> = Vec::with_capacity(selection.len());
+    let mut min_next_line = 0;
+    for selection in selection {
+        let (start, end) = selection.line_range(text);
+        let start = start.max(min_next_line).min(text.len_lines());
+        let end = (end + 1).min(text.len_lines());
+
+        lines.extend(start..end);
+        min_next_line = end;
+    }
+
+    let (line_commented, _, _, _) = find_line_comment(token.unwrap_or("//"), text, lines);
+    let comment_tokens = tokens.unwrap_or(("/*", "*/"));
+    let (block_commented, _) =
+        find_block_comments(comment_tokens.0, comment_tokens.1, text, selection);
+
+    // line comment when only line comment token available
+    if matches!(token, Some(_)) && !matches!(tokens, Some(_)) {
+        return true;
+        // block comment when only block comment tokens available
+    }
+    if !matches!(token, Some(_)) && matches!(tokens, Some(_)) {
+        return false;
+    }
```

Comment:
small nit: maybe move this in front of the comment search so that languages with just one type of comment avoid searching for the default comments. As  discussed in the matrix chat it would be nice to specical case languages with just block comments here, so returning an enum would be good. Also I think you can make this clearer as a  match:
``` rust 
let (token, tokens) = match (token, tokens){
 (Some(token), Some(tokens)) => (token, tokens),
 (None, None) => ("//", ("/*","*/")),
 (Some(_), None) => return CommentType::Line,
 (None, Some(_)) => return CommentType::BlockAsLineFallback,
};
```

Reply:
thanks for this review, I implemented it with some changes. The enum and match was a good call here. Turned out ``BlockAsLineFallback`` should only be returned if the selections are only one line long so I changed it to check for that(almost left this for tomorrow but I had an epiphany). I also didn't have the enum variants assigned a value because it wasn't needed and I didn't want to deal with lifetimes. Thank you again for this review it was very helpful.